dcap/types/src/quote3.rs

```diff
@@ -44,34 +77,217 @@ impl<T: AsRef<[u8]>> Quote3<T> {
+        Self::try_from_bytes(bytes)
     }
 }
 
-#[cfg(test)]
-mod test {
-    extern crate std;
+/// The Quoting enclave authentication data
+///
+/// Table 8 of
+/// <https://download.01.org/intel-sgx/latest/dcap-latest/linux/docs/Intel_SGX_ECDSA_QuoteLibReference_DCAP_API.pdf>.
+#[derive(Debug, Clone, Hash, PartialEq, Eq)]
+struct AuthenticationData<'a> {
+    bytes: &'a [u8],
+    // Since this has to be read always we unpack for availability.
+    // This is the `size` field, *not* the length of `bytes`.
+    data_size: usize,
+}
 
-    #[cfg(feature = "alloc")]
-    use std::vec;
+impl<'a> TryFrom<&'a [u8]> for AuthenticationData<'a> {
+    type Error = Error;
+    fn try_from(bytes: &'a [u8]) -> Result<Self> {
+        let bytes_length = bytes.len();
+
+        let data_size = u16_from_bytes(bytes, 0)? as usize;
+
+        let needed_size = data_size + mem::size_of::<u16>();
+        if bytes_length < needed_size {
+            Err(Error::InvalidInputLength {
+                required_size: needed_size,
+                actual_size: bytes_length,
+            })
+        } else {
+            Ok(Self { bytes, data_size })
+        }
+    }
+}
+
+impl<'a> AuthenticationData<'a> {
+    pub fn size(&self) -> usize {
+        self.data_size + mem::size_of::<u16>()
+    }
+}
+
+/// The Quoting enclave certification data
+///
+/// Table 9 of
+/// <https://download.01.org/intel-sgx/latest/dcap-latest/linux/docs/Intel_SGX_ECDSA_QuoteLibReference_DCAP_API.pdf>.
+#[derive(Debug, Clone, Hash, PartialEq, Eq)]
+struct CertificationData<'a> {
+    bytes: &'a [u8],
+    // Since these are small and the `size` has to be read we always unpack
+    // for availability.
+    data_type: u16,
+    // This is the size of the data field, *not* the size of `bytes`
+    data_size: usize,
+}
+
+impl<'a> TryFrom<&'a [u8]> for CertificationData<'a> {
+    type Error = Error;
+    fn try_from(bytes: &'a [u8]) -> Result<Self> {
+        let bytes_length = bytes.len();
+
+        // type (2 bytes) + size (4 bytes)
+        let mut needed_size = mem::size_of::<u16>() + mem::size_of::<u32>();
+
+        if bytes_length < needed_size {
+            return Err(Error::InvalidInputLength {
+                required_size: needed_size,
+                actual_size: bytes_length,
+            });
+        }
+
+        // These shouldn't fail since we ensured the length up above
+        let data_type = u16_from_bytes(bytes, 0)?;
+        let data_size = u32_from_bytes(bytes, mem::size_of::<u16>())? as usize;
+
+        needed_size += data_size;
+        if bytes_length < needed_size {
+            Err(Error::InvalidInputLength {
+                required_size: needed_size,
+                actual_size: bytes_length,
+            })
+        } else {
+            Ok(Self {
+                bytes,
+                data_type,
+                data_size,
+            })
+        }
+    }
+}
+
+fn u32_from_bytes(bytes: &[u8], offset: usize) -> Result<u32> {
```

Comment:
I'm a bit torn on these, they feel a bit indirect, but putting this logic inline in the `try_from()` methods felt very noisy.
I'm thinking maybe these should get moved into sgx utils and these get used in `core/types/src/report.rs ReportBody`

Reply:
I was unable to get a trait with implementation working in a repeatable fashion.
I think it's because there is no trait for the `from_le_bytes` function.
```rust
trait NumberFromSlice: Sized {
    const SIZE: usize;
    fn le_from_slice(bytes: &[u8]) -> Result<Self> {
        let value_bytes = bytes.get(..Self::SIZE).ok_or(Error::InputLength {
            required_size: Self::SIZE,
            actual_size: bytes.len(),
        })?;
        let mut copy_bytes = [0u8; Self::SIZE];
        copy_bytes.copy_from_slice(value_bytes);
        Ok(Self::from_le_bytes(copy_bytes))
    }
}
```shell
   Compiling mc-sgx-dcap-types v0.3.1-beta.0 (/home/nick/git/sgx/dcap/types)
error: constant expression depends on a generic parameter
   --> dcap/types/src/quote3.rs:261:36
    |
261 |         let mut copy_bytes = [0u8; Self::SIZE];
    |                                    ^^^^^^^^^^
    |
    = note: this may fail depending on what value the parameter takes

error[E0599]: no function or associated item named `from_le_bytes` found for type parameter `Self` in the current scope
   --> dcap/types/src/quote3.rs:263:18
    |
254 | trait NumberFromSlice: Sized {
    | ---------------------------- function or associated item `from_le_bytes` not found for this type parameter
...
263 |         Ok(Self::from_le_bytes(copy_bytes))
    |                  ^^^^^^^^^^^^^ function or associated item not found in `Self`

For more information about this error, try `rustc --explain E0599`.
error: could not compile `mc-sgx-dcap-types` due to 2 previous errors
```

I did however clean up the interface to pass in the slice and not have the offset